Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,194 @@
+        }
+    } else {
+
+        // new task version with auth support
+        var npmrcPath: string = path.join(cwd, '.npmrc');
+        var tempNpmrcPath : string = getTempNpmrcPath();
+        var debugLog: boolean = true;
+        if (tl.getVariable('system.debug') != undefined && tl.getVariable('system.debug').toLowerCase() === 'false') {
+            debugLog = false;
+        }
+
+        try{
+            await runNpmAuthHelperAsync(getNpmAuthHelperRunner(npmrcPath, tempNpmrcPath, debugLog));
+
+            // set required environment variables for npm execution
+            let npmExecOptions = <trm.IExecOptions>{
+                env: process.env
+            };
+            npmExecOptions.env['npm_config_userconfig'] = tempNpmrcPath;
+            if(debugLog) {
+                npmExecOptions.env['npm_config_loglevel'] =  'verbose';
+            }
+
+            await tryRunNpmConfigAsync(getNpmConfigRunner(debugLog), npmExecOptions);
+            var code : number =  await runNpmCommandAsync(npmRunner, npmExecOptions);
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    }
+}
+
+async function runNpmAuthHelperAsync(npmAuthRunner: trm.ToolRunner) : Promise<number> {
+    try{
+        var execOptions = <trm.IExecOptions>{
+            env: addBuildCredProviderEnv(process.env)
+        };
+
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Auth helper exitted with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        // ignore any auth failure and try to run the task.
+        tl.debug(tl.loc('NpmAuthFailed', err.message));
+    }
+}
+
+async function tryRunNpmConfigAsync(npmConfigRunner: trm.ToolRunner, execOptions : trm.IExecOptions) {
+    try {
+        var code = await npmConfigRunner.exec(execOptions);
+    } catch (err) {
+        // do not throw on this failure.
+        tl.warning(tl.loc('NpmConfigFailed', err.Message));
+    }
+}
+
+async function runNpmCommandAsync(npmCommandRunner: trm.ToolRunner, execOptions : trm.IExecOptions) : Promise<number> {
+    try{
+        var code: number = await npmCommandRunner.exec(execOptions);
+        tl.debug('Npm command succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmFailed', err.message));
+        throw err;
+    }
+}
+
+function shouldUseDeprecatedTask() : boolean {
+    return tl.getVariable('USE_DEPRECATED_TASK_VERSION') != undefined && tl.getVariable('USE_DEPRECATED_TASK_VERSION').toLowerCase() === 'true';
+}
+
+function getNpmAuthHelperRunner(npmrcPath: string, tempUserNpmrcPath: string, includeDebugLogs: boolean): trm.ToolRunner {
+    var npmAuthHelper = tl.tool(path.join(__dirname, 'Npm/vsts-npm-auth/bin/vsts-npm-auth.exe'));
```

Comment:
Right, for v1 atleast x-plat was not a consideration. we just don't want to break existing tasks. That's why our auth failure is non fatal. I still need to test in non windows platform though.